src/lib.rs

```diff
@@ -1,7 +1,372 @@
+    /// Create a [`WeightCache`] with a custom hasher.
+    pub fn with_hasher(capacity: NonZeroUsize, hasher: S) -> Self {
+        Self {
+            max: capacity.get(),
+            current: 0,
+            inner: LinkedHashMap::with_hasher(hasher),
+        }
+    }
+
+    /// Returns a reference to the value corresponding to the given key, if it
+    /// exists.
+    pub fn get(&mut self, k: &K) -> Option<&V> {
+        self.inner.get_refresh(k).map(|v| v as &V)
+    }
+
+    /// Returns a mutable reference to the value corresponding to the given key,
+    /// if it exists.
+    pub fn get_mut(&mut self, k: &K) -> Option<&mut V> {
+        self.inner.get_refresh(k)
+    }
+
+    /// Returns the number of key-value pairs in the cache.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns `true` if the cache contains no key-value pairs.
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Inserts a key-value pair into the cache. Returns an error if the value is
+    /// too bigger than the cache's configured max size.
+    pub fn put(&mut self, key: K, value: V) -> Result<(), ValueTooBigError> {
+        let weight = V::measure(&value);
+        if weight > self.max {
+            Err(ValueTooBigError)
+        } else {
+            self.current += weight;
+            // did we remove an element?
+            if let Some(x) = self.inner.insert(key, value) {
+                self.current -= V::measure(&x);
+            }
+
+            while self.current > self.max && !self.inner.is_empty() {
+                if let Some((_, v)) = self.inner.pop_front() {
+                    self.current -= V::measure(&v);
+                }
+            }
+
+            // remove elements until we're below the size boundary again
+            self.shrink_to_fit();
+            Ok(())
+        }
+    }
+
+    /// Returns an iterator over the cache's key-value pairs in least- to
+    /// most-recently-used order.
+    pub fn iter(&self) -> Iter<K, V> {
+        Iter(self.inner.iter())
+    }
+
+    /// Returns a mutable iterator over the cache's key-value pairs in least- to
+    /// most-recently-used order.
+    pub fn iter_mut(&mut self) -> IterMut<K, V> {
+        IterMut(self.inner.iter_mut())
+    }
+    fn shrink_to_fit(&mut self) {
+        while self.current > self.max && !self.inner.is_empty() {
+            let (_, v) = self.inner.pop_front().expect("Not empty");
+            self.current -= V::measure(&v);
+        }
+    }
+}
+impl<K: Hash + Eq, V, S: BuildHasher + Default> IntoIterator for WeightCache<K, V, S> {
+    type Item = (K, V);
+    type IntoIter = IntoIter<K, V>;
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter(self.inner.into_iter())
+    }
+}
+/// An iterator over a cache's key-value pairs in least- to most-recently-used
+/// order.
+#[derive(Clone)]
+pub struct IntoIter<K, V>(linked_hash_map::IntoIter<K, V>);
+
+impl<K, V> Iterator for IntoIter<K, V> {
```

Comment:
Do we actually need all these iterator structs, or are they just for completeness sake?

Reply:
See my other comment. We can also just remove it until we need it.